### PR TITLE
Adapt throw test for Promise-returning operations

### PR DESCRIPTION
**Pull request by dontcallmedom**

### CODE REVIEW

**Comment by zqzhang** [comment]
Seems idlharness.js based battery status tests got one more failure with this change, on Chrome Beta for Android.

- http://zqzhang.github.io/demo/idlharness/battery-status.html with this change applied.
- http://w3c-test.org/battery-status/battery-interface-idlharness.html without.

File: idlharness.js [hunk 1]
```diff
@@ -69,6 +69,22 @@ function minOverloadLength(overloads) {
...
 }
 
+function throwOrReject(operation, fn, obj, args,  message) {
+    if (operation.idlType.generic !== "Promise") {
+        assert_throws(new TypeError(), function() {
+            fn.apply(obj, args);
+        }, message);
+    } else {
+        fn.apply(obj, args).then(
+                function() {
```
Comment (zqzhang) [inline idlharness.js]: No need 4 more space indent for this code block, from the coding style of the whole file.

**Comment by Ms2ger** [comment]
Using `promise_rejects` would be better if we can thread the test object through.

File: idlharness.js (hunk 1 above)
Comment (dontcallmedom) [inline idlharness.js]: fixed

**Comment by dontcallmedom** [comment]
@Ms2ger it now uses `promise_rejects`

**Comment by Ms2ger** [comment]
I don't think it makes sense to create a new `async_test`. We should ensure that the failure is reported on the same test we'd use for the non-promise case.

**Comment by dontcallmedom** [comment]
I don't think idlharness uses asynchronous tests otherwise, since it general, it tests stuff synchronously.

**Comment by zqzhang** [comment]
> I don't think it makes sense to create a new async_test. We should ensure that the failure is reported on the same test we'd use for the non-promise case.

+1

It's not acceptable to report a failure in a separate test case.

**Comment by dontcallmedom** [comment]
@Ms2ger @zqzhang I have updated the pull request, can you check whether it matches what you requested?

**Comment by dontcallmedom** [comment]
@Ms2ger @zqzhang I have updated the pull request, can you check whether it matches what you requested?

File: idlharness.js [hunk 2]
```diff
@@ -1048,9 +1076,13 @@ IdlInterface.prototype.test_member_attribute = function(member)
 IdlInterface.prototype.test_member_operation = function(member)
 //@{
 {
-    test(function()
+    var a_test = async_test(this.name + " interface: operation " + member.name +
+    "(" + member.arguments.map(function(m) { return m.idlType.idlType; }) +
```
Comment (zqzhang) [inline idlharness.js]: Proper indent please, and next line.

File: idlharness.js [hunk 3]
```diff
@@ -1142,12 +1173,14 @@ IdlInterface.prototype.do_member_operation_asserts = function(memberHolderObject
...
     if (!member["static"]) {
+        var cb;
         if (!this.is_global() &&
             memberHolderObject[member.name] != self[member.name])
         {
-            assert_throws(new TypeError(), function() {
-                memberHolderObject[member.name].apply(null, args);
-            }, "calling operation with this = null didn't throw TypeError");
+            cb = awaitNCallbacks(2, done);
+            throwOrReject(a_test, member, memberHolderObject[member.name], null, args,  "calling operation with this = null didn't throw TypeError", cb);
```
Comment (zqzhang) [inline idlharness.js]: Better with a break before message.

File: idlharness.js [hunk 4]
```diff
@@ -1155,9 +1188,9 @@ IdlInterface.prototype.do_member_operation_asserts = function(memberHolderObject
         //
         // TODO: Test a platform object that implements some other
         // interface.  (Have to be sure to get inheritance right.)
-        assert_throws(new TypeError(), function() {
-            memberHolderObject[member.name].apply({}, args);
-        }, "calling operation with this = {} didn't throw TypeError");
+        throwOrReject(a_test, member, memberHolderObject[member.name], {}, args,  "calling operation with this = {} didn't throw TypeError", cb);
```
Comment (zqzhang) [inline idlharness.js]: Better with a break before message.

**Comment by zqzhang** [comment]
LGTM with nits.

**Pull request merged**